spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,156 @@
+      will be violated, unless positional use is deprecated. Specifically:
+
+      - If `None` is passed by position, the function will change from being
+        seeded to being unseeded.
+      - If an integer is passed by position, the random stream will change.
+
+      We suggest that projects consider deprecating positional use of
+      `random_state`/`rng` (i.e., change their function signatures to
+      ``def my_func(..., *, rng=None)``); that might not make sense
+      for all projects, so this SPEC does not make that
+      recommendation, neither does this decorator enforce it.
+
+    Parameters
+    ----------
+    old_name : str
+        The old name of the PRNG argument (e.g. `seed` or `random_state`).
+    position_num : int, optional
+        The (0-indexed) position of the old PRNG argument (if accepted by position).
+        Maintainers are welcome to eliminate this argument and use, for example,
+        `inspect`, if preferred.
+    dep_version : str, optional
+        The full version number of the library in which warnings about use of the
+        old PRNG argument begin to be emitted. The version in which the behavior
+        will change is assumed to be two versions later.
+
+    """
+    NEW_NAME = "rng"
+
+    if dep_version:
+        end_version = dep_version.split(".")
+        end_version[1] = str(int(end_version[1]) + 2)
+        end_version = ".".join(end_version)
+
+    def decorator(fun):
+        @functools.wraps(fun)
+        def wrapper(*args, **kwargs):
+            # Determine how PRNG was passed
+            as_old_kwarg = old_name in kwargs
+            as_new_kwarg = NEW_NAME in kwargs
+            as_pos_arg = position_num is not None and len(args) >= position_num + 1
+
+            # Can only specify PRNG one of the three ways
+            if int(as_old_kwarg) + int(as_new_kwarg) + int(as_pos_arg) > 1:
+                message = (
+                    f"{fun.__name__}() got multiple values for "
+                    f"argument now known as `{NEW_NAME}`"
+                )
+                raise TypeError(message)
+
+            cmn_msg = (
+                " To silence this warning and ensure consistent behavior in "
+                f"SciPy {end_version}, control the RNG using argument "
+                f"`{NEW_NAME}`. Arguments passed to keyword `{NEW_NAME}` will "
+                "be validated by `np.random.default_rng`, so the behavior "
+                "corresponding with a given value may change compared to use "
+                f"of `{old_name}`. For example, "
+                "1) `None` results in unpredictable random numbers, "
+                "2) an integer results in a different stream of random numbers, "
+                "(with the same distribution), and "
+                "3) `np.random` or `RandomState` instances result in an error. "
+                "See the documentation of `default_rng` for more information."
+            )
```

Comment:
It probably doesn't matter as it is simple, but you could interpolate this string in the outer.  Because then you don't have to do the work unless you give the warning.

Reply:
"in the outer", meaning before `def decorator`? 

> Because then you don't have to do the work unless you give the warning.

Wouldn't it always do the work if there is a decorated function, but only once?

Reply:
Yes, before the `def wrapper` would also work.  Anyway, was just a small nit (this is a problem mainly in tests where you print arrays inside these sometimes).

Reply:
Addressed by https://github.com/stefanv/specs/pull/3/commits/4aa7051ee73a82bb676ddfab1d4aa310c1f9267e